pkg/auth/providers/migration/activedirectory_migration.go

```diff
@@ -0,0 +1,243 @@
+package migration
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/auth/providers"
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	crtbsByPrincipalAndUserIndex = "auth.management.cattle.io/crtbByPrincipalAndUser"
+	prtbsByPrincipalAndUserIndex = "auth.management.cattle.io/prtbByPrincipalAndUser"
+	tokenByUserRefKey            = "auth.management.cattle.io/token-by-user-ref"
+)
+
+type adMigration struct {
+	users        v3.UserInterface
+	userLister   v3.UserLister
+	tokenLister  v3.TokenLister
+	tokens       v3.TokenInterface
+	prtbs        v3.ProjectRoleTemplateBindingInterface
+	crtbs        v3.ClusterRoleTemplateBindingInterface
+	crtbIndexer  cache.Indexer
+	prtbIndexer  cache.Indexer
+	tokenIndexer cache.Indexer
+	authConfigs  v3.AuthConfigInterface
```

Comment:
Looks like you're not using authConfigs, secrets, certs, caPool, and tokenLister.